docs/src/main/asciidoc/security-openid-connect-multitenancy-concept.adoc

```diff
@@ -0,0 +1,317 @@
+
+With every call, the multi-tenant application can translate the `tenant identifier` "org1" to imply that the user is logged in to the SSO of organisation 1, and "org2" to the SSO of organisation 2. The multi-tenant application can also verify that the user information in the `IdToken` is not changed, and extract the data within it.
+
+
+== The 20-minute overview of a multi-tenant application
+
+Quarkus multi-tenant applications need to perform several tasks with each call from a client to provide a simple and consistent development experience. Understanding the sequence of those steps helps designing multi-tenant applications correctly; the conceptual models described in the OIDC and OAuth2 specifications is huge and illustrating the steps required by development teams helps to clarify multi-tenant application development.
+
+In this example, our multi-tenant application is designed to serve RESTful resources to already logged-in users. This mode of operation is called `Bearer authentication` in OIDC and indicates that an HTTP header called `"authorization:"` with the value `"Bearer *encoded JWT token*"` must be present in each request from the client to the application.
+
+.Partial steps of a multi-tenant application
+[caption="Figure 2: "]
+image::oidc-multitenancy-concept-2.svg[Multitenant Application Steps]
+
+The illustration below is somewhat simplified, but covers the main lifecycle security tasks within a multi-tenant application. Figure 2 illustrates a multi-tenant application which receives a call from a user authenticated within organisation 2 and the tasks performed by Quarkus shown in blue squares. The 3 numbered, red circles indicate process steps where you need to supply custom implementations to ensure that the multi-tenant application works well.
+
+The 3 required steps are detailed below, along with descriptions of concepts and some considerations:
+
+. Find `tenant identifier` from request, and lookup/resolve the corresponding OIDC configuration for the active tenant.
+. Inject user information (packaged in a `JsonWebToken` object) from the active `AccessToken`. This includes implementing reusable logic to simplify access to information given by OIDC.
+. Validate user authorization to retrieve or create data.
+
+=== 1. Resolve the `tenant identifier` and `OidcTenantConfig`
+
+Before being able to execute the actual business logic of a multi-tenant application, Quarkus needs to enforce security requirements for authentication and authorization. To do that, Quarkus must verify that received `JsonWebTokens` are valid, and that the information within them permits the active user to read or update the current resource. In order to do so, Quarkus must read the active `tenant id` of the current request and use it to resolve the active tenant's OIDC configuration (defined by an instance of `OidcTenantConfig`). Finding the tenant id from the current request can be done in one of 2 ways:
```

Comment:
```suggestion
Before being able to execute the actual business logic of a multi-tenant application, Quarkus needs to enforce security requirements for authentication and authorization. To do that, Quarkus must verify that received ID and/or access tokens are valid, and that the information within them permits the active user to read or update the current resource. In order to do so, Quarkus must read the active `tenant id` of the current request and use it to resolve the active tenant's OIDC configuration (defined by an instance of `OidcTenantConfig`). Finding the tenant id from the current request can be done in one of 2 ways:
```

Reply:
This is an attempt to reflect that 1) bearer tokens can be in JWT or opaque format 2) in the code flow it is typically ID token, immediately or as a session cookie, is verified - and the code flow access token can also be verified